### ISSUE TITLE
Seqr loading pipeline workflow

### ISSUE THREAD

**Issue by mike-w-wilson** [broadinstitute/seqr#3042]
1. After the PMs alert the #exome_wgs_callset channel of a new callset, I navigate to the [RDG tracker airtable](https://airtable.com/app3Y97xtbbaOopVR/tblG1k2YLBu4bw2XY/viwW0oZeRVcbgTduD?blocks=hide) to see which projects are labeled "Methods (pre-relatedness" 
2. I open up the seqr link for each of these projects and copy the project guid to a txt file that will be used in the [pedigree pull script](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/seqr_scripts/pull_seqr_pedigrees.py)
3. I then click on the Terra workspace associated with the new projects which brings me to the workspaces "data" page. I navigate to the files in the left hand toolbar and then find the most recent callset directory. This directories structure is not always the same and you may have to click into another subdirectory to find the VCF 🕵️. WGS callsets are a single sharded VCF. WES callsets are delivered with two outputs of sharded VCF files that need to be merged and we use this [script](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/seqr_scripts/import_vcf.py). I write the merged file to the seqr-temp bucket. Stacey asked DSP and they said eventually they want to move WES to the same pipeline as WGS so the merge should eventually disappear. 
4. For WGS callsets, I create a remap file for the new VCFs since DSP changes the sample IDs. This is my current sed/awk noob approach which is hardcoded and  is ugly but it works. I update this by version and path.
```
gsutil cat  gs://fc-01aa37cb-3fe1-4723-b498-6c3045a50ac0/RDG_Broad_WGS_Internal_All_Aug2022/v2/0000000000-RDG-Broad-WGS-Internal.vcf.gz | zgrep "#CHROM" > v32_samples.tsv
cut -f10- v32_samples.tsv | sed 's/\t/\n/g' > remap_callset.txt
cut -f10- v32_samples.tsv | sed 's/\t/\n/g' | sed 's/_[^_]*$//g' | sed 's/_[^_]*$//g'|sed 's/_[^_]*$//g' | sed 's/^[^_]*_//' > v32_seqr_ids.txt
paste remap_callset.txt v32_seqr_ids.txt > remap.tsv
```
I then copy this file to the bucket of the current callset version so something like 
```
gsutil cp remap.tsv gs://seqr-datasets/v02/GRCh38/RDG_WGS_Broad_Internal/v32/RDG_WGS_Broad_Internal_remap.tsv
```
This is where airflow expects the file to be if you are trying to remap the vcf.
6. Next in [airflow variables](https://rf1b81cf3e6c1c7e7p-tp.appspot.com/variable/list/), I update the appropriate DAG's variables. I enter the `vcf_path` (for WGS its the terra bucket vcf path, WES is the merged seqr-temp vcf path) and update the `version_path` to whichever version we are on. I then clear all projects from `projects_to_run` so it is just an empty list  so only the first step runs because I use it as the input to the relatedness checks. The relatedness checks do not need this step to actually run, we could start with the VCF, but its sorta a two birds with one stone situation. Then I trigger the DAG. 
7. While the first step runs, I run the pull_seqr_pedigrees.py script with the file created in step 2. This requires some the classic psuedohack of "Inspect Element" and then endlessly click on all tabs and tables until I find  where my session_id is listed because I forgot where it is every time. 
8. The first step takes a while so I also usually generate any remap file a project may need. To do this, I view the VCF IDs and SeqrIDs column in airtable. If a project has entries in those fields, I check to see if an existing remap file is in the projects base bucket directory e.g. gs://seqr-datasets/v02/GRCh38/RDG_WGS_Broad_Internal/base/projects/R0384_rare_genomes_project_gen/R0384_rare_genomes_project_gen_remap.tsv and download that file with gsutil. I then add the new samples that need remapping from airtable(assume the two fields are lists that can be zipped) and reupload the file. This will be used in the second step of the pipeline but do it now since step one takes hours.
9. Once the first step of the loading pipeline completes, I spin up a dataproc cluster, usually a 2-worker n1-standard-8 machine with a 20 secondary worker autoscaling policy and I run the [sex_check.py](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/seqr_scripts/sex_check.py) script. Once the sex check script completes, I run the [relatedness script](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/relatedness/get_ibd.py). For more information on how these scripts are run see [this section](https://docs.google.com/document/d/1gORkEEVy026robiuAaLdaeJM10OzTJmppP9chDihNuE/edit#heading=h.n9u1qqwsn9nu) in the loading docs.
10. I then pass the relatedness scripts outputs (one file per project and the sex check file) to the PMs in slack like [so](https://the-tgg.slack.com/archives/C0L6QDZ70/p1667391527788779?thread_ts=1666625243.822209&cid=C0L6QDZ70) and I update the projects' status in Airtable to "PM Team (Relatedness Checks)".
11. The PMs make necessary adjustments in Airtable for samples that will not be loaded and then update me in slack that the checks are complete. 
12. I then copy the project guids out of the pedigree file created in step 2 and  go back to the airflow variable page and paste the projects into the callset's DAG variable `projects_to_run` and format them correctly -- quotes on each project in the list, indented six spaces, and no comma after the last project entry. I then trigger the dag.
13. Once a project is successfully exported, I navigate to the project page in seqr, usually through the Airtable and open another tab on https://seqr.broadinstitute.org/data_management/elasticsearch_status and copy and paste the index into the project. 
14. I then tag the analyst listed in airtable and the PM team on  the slack notification that new samples were added to the project. 
15. Lastly, I update the status in airtable to "PM Team(crams)" .

I think this covers the workflow. If you have any questions, let me know.

**Comment by hanars** [comment]
> 1. After the PMs alert the #exome_wgs_callset channel of a new callset, I navigate to the [RDG tracker airtable](https://airtable.com/app3Y97xtbbaOopVR/tblG1k2YLBu4bw2XY/viwW0oZeRVcbgTduD?blocks=hide) to see which projects are labeled "Methods (pre-relatedness"

In seqr UI, PMs will select projects

> 2. I open up the seqr link for each of these projects and copy the project guid to a txt file that will be used in the [pedigree pull script](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/seqr_scripts/pull_seqr_pedigrees.py)

No longer needed

> 3. I then click on the Terra workspace associated with the new projects which brings me to the workspaces "data" page. I navigate to the files in the left hand toolbar and then find the most recent callset directory. This directories structure is not always the same and you may have to click into another subdirectory to find the VCF 🕵️. 

In seqr UI, PMs will select workspace/vcf directory

> WGS callsets are a single sharded VCF. WES callsets are delivered with two outputs of sharded VCF files that need to be merged and we use this [script](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/seqr_scripts/import_vcf.py). I write the merged file to the seqr-temp bucket. Stacey asked DSP and they said eventually they want to move WES to the same pipeline as WGS so the merge should eventually disappear.

Add to airflow

> 4. For WGS callsets, I create a remap file for the new VCFs since DSP changes the sample IDs. This is my current sed/awk noob approach which is hardcoded and  is ugly but it works. I update this by version and path.
> ...
> I then copy this file to the bucket of the current callset version so something like
> ...
> This is where airflow expects the file to be if you are trying to remap the vcf. 6.

Add to seqr loading endpoint

> 6. Next in [airflow variables](https://rf1b81cf3e6c1c7e7p-tp.appspot.com/variable/list/), I update the appropriate DAG's variables. I enter the `vcf_path` (for WGS its the terra bucket vcf path, WES is the merged seqr-temp vcf path) and update the `version_path` to whichever version we are on... Then I trigger the DAG. 

Add to seqr loading endpoint

> I then clear all projects from `projects_to_run` so it is just an empty list so only the first step runs because I use it as the input to the relatedness checks. The relatedness checks do not need this step to actually run, we could start with the VCF, but its sorta a two birds with one stone situation. 

Will not be needed anymore

> 7.  While the first step runs, I run the pull_seqr_pedigrees.py script with the file created in step 2. This requires some the classic psuedohack of "Inspect Element" and then endlessly click on all tabs and tables until I find where my session_id is listed because I forgot where it is every time. 

Add to seqr loading endpoint

> 8. The first step takes a while so I also usually generate any remap file a project may need. To do this, I view the VCF IDs and SeqrIDs column in airtable. If a project has entries in those fields, I check to see if an existing remap file is in the projects base bucket directory e.g. gs://seqr-datasets/v02/GRCh38/RDG_WGS_Broad_Internal/base/projects/R0384_rare_genomes_project_gen/R0384_rare_genomes_project_gen_remap.tsv and download that file with gsutil. I then add the new samples that need remapping from airtable(assume the two fields are lists that can be zipped) and reupload the file. This will be used in the second step of the pipeline but do it now since step one takes hours. 

Add to airflow

> 9. Once the first step of the loading pipeline completes, I spin up a dataproc cluster, usually a 2-worker n1-standard-8 machine with a 20 secondary worker autoscaling policy and I run the [sex_check.py](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/seqr_scripts/sex_check.py) script. Once the sex check script completes, I run the [relatedness script](https://github.com/broadinstitute/tgg_methods/blob/master/tgg/relatedness/get_ibd.py). For more information on how these scripts are run see [this section](https://docs.google.com/document/d/1gORkEEVy026robiuAaLdaeJM10OzTJmppP9chDihNuE/edit#heading=h.n9u1qqwsn9nu) in the loading docs. 

Add to airflow

> 10. I then pass the relatedness scripts outputs (one file per project and the sex check file) to the PMs in slack like [so](https://the-tgg.slack.com/archives/C0L6QDZ70/p1667391527788779?thread_ts=1666625243.822209&cid=C0L6QDZ70) and I update the projects' status in Airtable to "PM Team (Relatedness Checks)".

Add to airflow. Will update the status in airtble and alert the PMS. Will happen before the compute subset step so loading will proceed without failed samples

> 11. The PMs make necessary adjustments in Airtable for samples that will not be loaded and then update me in slack that the checks are complete. 

Will not be needed, will have to re-trigger loading on a project later if a sample now passes relatedness

> 12. I then copy the project guids out of the pedigree file created in step 2 and go back to the airflow variable page and paste the projects into the callset's DAG variable `projects_to_run` and format them correctly -- quotes on each project in the list, indented six spaces, and no comma after the last project entry. I then trigger the dag. 

Will set this in the original DAG trigger in the seqr loading endpoint

> 13. Once a project is successfully exported, I navigate to the project page in seqr, usually through the Airtable and open another tab on https://seqr.broadinstitute.org/data_management/elasticsearch_status and copy and paste the index into the project. 

Will keep this manual for now, with hail backend will make sure this is automated

> 14. I then tag the analyst listed in airtable and the PM team on the slack notification that new samples were added to the project. 

This can be added to seqr, independent of any other loading changes

> 15. Lastly, I update the status in airtable to "PM Team(crams)" .

This can be added to seqr, independent of any other loading changes

**Comment by hanars** [comment]
Airflow changes:

1. Script to merge WES VCFs
2. Create project specific remap file based on airtable SeqrID vs VCF ID
3. Sex check and relatedness check. Will modify airtable and update PMs on failures


For seqr, once that is done we can build a UI to trigger the DAG once. The last two manual steps for notifying people can beautomated as part of a separate ticket
